File: java-samples/referencestates-sanctionsbody/workflows/src/main/java/com/r3/developers/samples/referencestate/workflows/IssueSanctionsListFlow.java
package com.r3.developers.samples.referencestate.workflows;

import com.r3.developers.samples.referencestate.contracts.SanctionListContract;
import com.r3.developers.samples.referencestate.states.Member;
import com.r3.developers.samples.referencestate.states.SanctionList;
import net.corda.v5.application.flows.ClientRequestBody;
import net.corda.v5.application.flows.ClientStartableFlow;
import net.corda.v5.application.flows.CordaInject;
import net.corda.v5.application.flows.InitiatingFlow;
import net.corda.v5.application.membership.MemberLookup;
import net.corda.v5.application.messaging.FlowMessaging;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.base.exceptions.CordaRuntimeException;
import net.corda.v5.ledger.common.NotaryLookup;
import net.corda.v5.ledger.utxo.UtxoLedgerService;
import net.corda.v5.ledger.utxo.transaction.UtxoSignedTransaction;
import net.corda.v5.ledger.utxo.transaction.UtxoTransactionBuilder;
import net.corda.v5.membership.MemberInfo;
import net.corda.v5.membership.NotaryInfo;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * This flows allows a party to issue a sanctions list.
 * This sanctions list will be used by other parties when they are making their
 * IOU agreements to determine whether the counter-party is trustworthy.
 *
 */
@InitiatingFlow(protocol = "issue-sanction-list")
public class IssueSanctionsListFlow implements ClientStartableFlow {

    private final static Logger log = LoggerFactory.getLogger(IssueSanctionsListFlow.class);
    @CordaInject
    private MemberLookup memberLookup;
    @CordaInject
    private NotaryLookup notaryLookup;
    @CordaInject
    private UtxoLedgerService ledgerService;
    @CordaInject
    private FlowMessaging flowMessaging;

    @NotNull
    @Override
    @Suspendable
    public String call(@NotNull ClientRequestBody requestBody) {
        try{
            MemberInfo myInfo = memberLookup.myInfo();
            NotaryInfo notary = notaryLookup.getNotaryServices().iterator().next();
            List<MemberInfo> allParties = memberLookup.lookup();
            allParties = allParties.stream().filter(it ->
                    !it.getName().getCommonName().contains("Notary")).collect(Collectors.toList());

            SanctionList state = new SanctionList(
                    Collections.emptyList(),
                    new Member(myInfo.getName(), myInfo.getLedgerKeys().get(0)),
                    allParties.stream().map(it->it.getLedgerKeys().get(0)).collect(Collectors.toList())
            );
            UtxoTransactionBuilder transactionBuilder = ledgerService.createTransactionBuilder()
                    .setNotary(notary.getName())
                    .setTimeWindowBetween(Instant.now(), Instant.now().plusMillis(Duration.ofDays(1).toMillis()))
                    .addOutputState(state)
                    .addCommand(new SanctionListContract.Create())
                    .addSignatories(myInfo.getLedgerKeys().get(0));

            UtxoSignedTransaction signedTransaction = transactionBuilder.toSignedTransaction();

            List<MemberInfo> otherParties = new ArrayList<>(allParties);
            otherParties.remove(myInfo);

            UtxoSignedTransaction finalizedSignedTransaction =
                    ledgerService.finalize(signedTransaction,
                            otherParties.stream().map(
                                    it->flowMessaging.initiateFlow(it.getName())).collect(Collectors.toList())
                            ).getTransaction();

            return finalizedSignedTransaction.getId().toString();

        }catch (Exception e){
            log.warn("Failed to process request because: " + e.getMessage());
            throw new CordaRuntimeException(e.getMessage());
        }
    }
}

/* Example JSON to put into REST-API POST requestBody
 {
    "clientRequestId": "issue-sanction",
    "flowClassName": "com.r3.developers.samples.referencestate.workflows.IssueSanctionsListFlow",
    "requestBody": {
    }
 }
 */

File: kotlin-samples/referencestates-sanctionsbody/workflows/src/main/kotlin/com/r3/developers/samples/referencestate/workflows/IssueSanctionsListFlow.kt
package com.r3.developers.samples.referencestate.workflows

import com.r3.developers.samples.referencestate.contracts.SanctionListContract
import com.r3.developers.samples.referencestate.states.Member
import com.r3.developers.samples.referencestate.states.SanctionList
import net.corda.v5.application.flows.*
import net.corda.v5.application.membership.MemberLookup
import net.corda.v5.application.messaging.FlowMessaging
import net.corda.v5.application.messaging.FlowSession
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.ledger.common.NotaryLookup
import net.corda.v5.ledger.utxo.UtxoLedgerService
import net.corda.v5.ledger.utxo.transaction.UtxoSignedTransaction
import net.corda.v5.membership.MemberInfo
import org.slf4j.LoggerFactory
import java.time.Duration
import java.time.Instant
import kotlin.streams.toList


@InitiatingFlow(protocol = "issue-sanction-list")
class IssueSanctionsListFlow : ClientStartableFlow {

    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }

    @CordaInject
    lateinit var memberLookup: MemberLookup

    // Injects the UtxoLedgerService to enable the flow to make use of the Ledger API.
    @CordaInject
    lateinit var ledgerService: UtxoLedgerService

    @CordaInject
    lateinit var notaryLookup: NotaryLookup

    @CordaInject
    lateinit var flowMessaging: FlowMessaging

    @Suspendable
    override fun call(requestBody: ClientRequestBody): String {
        try {
            val myInfo = memberLookup.myInfo()
            val notary = notaryLookup.notaryServices.single()
            val allParties = memberLookup.lookup().filter { it -> !it.name.commonName.equals("NotaryRep1") }.toList()

            val sanctionList = SanctionList(
                badPeople = emptyList(),
                issuer = Member(myInfo.name, myInfo.ledgerKeys[0]),
                participants = allParties.map { it.ledgerKeys[0] }.toList()
            )

            val txBuilder = ledgerService.createTransactionBuilder()
                .setNotary(notary.name)
                .setTimeWindowBetween(Instant.now(), Instant.now().plusMillis(Duration.ofDays(1).toMillis()))
                .addOutputState(sanctionList)
                .addCommand(SanctionListContract.SanctionListCommand.Create())
                .addSignatories(myInfo.ledgerKeys[0])

            val signedTransaction: UtxoSignedTransaction = txBuilder.toSignedTransaction()

            val finalizedSignedTransaction = ledgerService.finalize(
                signedTransaction,
                (allParties - myInfo).stream().map { it: MemberInfo -> flowMessaging.initiateFlow(it.name) }.toList()
            ).transaction

            return finalizedSignedTransaction.id.toString()
        }catch (e: Exception) {
            log.warn("Failed to process utxo flow for request body '$requestBody' because:'${e.message}'")
            throw e
        }
    }
}

@InitiatedBy(protocol = "issue-sanction-list")
class IssueSanctionsListFlowResponder : ResponderFlow {

    private companion object {
        val log = LoggerFactory.getLogger(this::class.java.enclosingClass)
    }

    @CordaInject
    lateinit var ledgerService: UtxoLedgerService

    @Suspendable
    override fun call(session: FlowSession) {
        log.info("IssueSanctionsListFlowResponder.call() called")

        try {
            val finalizedSignedTransaction = ledgerService.receiveFinality(session) { ledgerTransaction -> }
            log.info("Finished responder flow - ${finalizedSignedTransaction.transaction.id}")
        }
        // Soft fails the flow and log the exception.
        catch (e: Exception) {
            log.warn("Exceptionally finished responder flow", e)
        }

    }
}